YarnSpinner/VirtualMachine.cs

```diff
@@ -117,60 +116,209 @@ internal enum TokenType {
+                            break;
+                        case Yarn.Value.Type.Number:
+                            this._yarnValue = new Yarn.Value(float.Parse(this.Value));
+                            break;
+                        case Yarn.Value.Type.String:
+                            this._yarnValue = new Yarn.Value(this.Value);
+                            break;
+                        case Yarn.Value.Type.Bool:
+                            this._yarnValue = new Yarn.Value(bool.Parse(this.Value));
+                            break;
+                    }
                 }
+
+                return this._yarnValue;
             }
+            private set {
+                this._yarnValue = value;
+            }
+        }
+    }
 
-            /// Pop a value from the stack
-            public Value PopValue() {
-                return stack.Pop ();
+    public partial class VMState
+    {
+        // The methods in here used to handle a Stack object, but it is easier to serialize
+        // and handle a List these methods are now providing Stack functionality on top of a List.
+
+        /// Methods for working with the stack
+        public void PushValue(object o) {
+            if (o is Value) {
+                Stack.Add(new VMStateValue(o as Value));
+            } else {
+                Stack.Add(new VMStateValue(new Value(o)));
             }
+        }
 
-            /// Peek at a value from the stack
-            public Value PeekValue() {
-                return stack.Peek ();
+        /// Pop a value from the stack
+        public Value PopValue() {
+            if (Stack.Count != 0) {
+                int lastElement = Stack.Count - 1;
+                Value item = Stack[lastElement].YarnValue;
+                Stack.RemoveAt(lastElement);
+
+                return item;
             }
 
-            /// Clear the stack
-            public void ClearStack() {
-                stack.Clear ();
+            return null;
+        }
+
+        /// Peek at a value from the stack
+        public Value PeekValue() {
+            if (Stack.Count != 0) {
+                return Stack[Stack.Count - 1].YarnValue;
             }
+
+            return null;
+        }
+
+        /// Clear the stack
+        public void ClearStack() {
+            Stack.Clear();
         }
-        
+    }
+
+    internal class VirtualMachine
+    {
         internal VirtualMachine (Dialogue d)
         {
             dialogue = d;
-            state = new State ();
+            state = new VMState();
         }
 
         /// Reset the state of the VM
         internal void ResetState() {
-            state = new State();
+            state = new VMState();
+        }
+
+        // Using this flag to return expected VM State programCounter.
+        // depends on when the State is requested, mostly it would be after
+        // an instruction has been processed which increments the counter.
+        private bool instructionProcessed = false;
```

Comment:
I am slightly confused as to the purpose of this flag.
It appears to switch between on and off based on how far through running an instruction you are but so is it for solving the issue of when you want to save the state?

If that is the case I don't really think that specifically is worth supporting, you have control over when you save and load the state so handling edge cases like that just feels like a debugging nightmare waiting to happen.

Or have I misunderstood its purpose?
Could you please add some more information as to this flags purpose?

Reply:
It is something I came up with while testing the solution, at the moment the only way to get correct programCounter is to call GetState before instruction is processed. Which is through UIBehaviour events like OnLineStart, OnOptionsStart etc. 

If I am just starting at the screen at a dialogue currently shown on the screen and press save button or use menu to save my state, the programCounter at that point will be +1 while stack and values will be same.

Loading such state breaks the VM.

I haven't thought of better solution but will try to improve it.

I tried moving programCounter++ in continue method before the RunInstruction() but that did break some code as well so it might need more work.